src/Language/DifferentialDatalog/TypeInference.hs

```diff
@@ -0,0 +1,637 @@
+
+deIsFloat_ :: DDExpr -> Predicate
+deIsFloat_ de = PEq (TETypeOfExpr de) TEFloat
+
+deIsDouble_ :: DDExpr -> Predicate
+deIsDouble_ de = PEq (TETypeOfExpr de) TEDouble
+
+deIsFP :: DDExpr -> Constraint
+deIsFP de = CDisj $ deIsFP_ de
+
+deIsFP_ :: DDExpr -> [[Predicate]]
+deIsFP_ de = [[deIsFloat_ de], [deIsDouble_ de]]
+
+-- 'is_bits t = is_Bit t || is_Signed t'.
+deIsBits :: DDExpr -> Constraint
+deIsBits de = CDisj $ deIsBits_ de
+
+deIsBits_ :: DDExpr -> [[Predicate]]
+deIsBits_ de = [[deIsBit_ de], [deIsSigned_ de]]
+
+-- 'is_int t = is_bits t || is_BigInt t'.
+deIsInt :: DDExpr -> Constraint
+deIsInt de = CDisj $ deIsInt_ de
+
+deIsInt_ :: DDExpr -> [[Predicate]]
+deIsInt_ de = deIsBits_ de ++ [[deIsBigInt_ de]]
+
+-- 'is_num t = is_int t || is_fp t'.
+deIsNum :: DDExpr -> Constraint
+deIsNum de = CDisj $ deIsInt_ de ++ deIsFP_ de
+
+-- Type expression is a shared reference type with specified inner type.
+deIsSharedRef :: (?d::DatalogProgram) => DDExpr -> TExpr -> [Constraint]
+deIsSharedRef de t =
+    cdisj $ map (\tdef -> [PEq (TETypeOfExpr de) (TEExtern (name tdef) [t])]) sref_tdefs
+    where
+    -- All shared reference types.
+    sref_tdefs = filter (tdefGetSharedRefAttr ?d)
+                 $ M.elems $ progTypedefs ?d
+
+-- Expression is a set (Set, Tinyset, Vec, ..) with specified element type.
+deIsSet :: (?d::DatalogProgram) => DDExpr -> TExpr -> Constraint
+deIsSet de t = CDisj $ map (deIsSet_ de t) sET_TYPES
+
+-- Expression is a collection with specified element type (when iterating using
+-- for-loop).
+deIsIterable :: (?d::DatalogProgram) => DDExpr -> TExpr -> Constraint
+deIsIterable de t =
+    CDisj $
+    deIsGroup_ de t :
+    deIsMap_ de t :
+    map (deIsSet_ de t) sET_TYPES
+
+deIsSet_ :: (?d::DatalogProgram) => DDExpr -> TExpr -> String -> [Predicate]
+deIsSet_ de t set_type =
+    [PEq (TETypeOfExpr de) (TEExtern set_type [t])]
+    where
+    TypeDef{tdefArgs=[_],..} = getType ?d set_type
+
+deIsGroup_ :: (?d::DatalogProgram) => DDExpr -> TExpr -> [Predicate]
+deIsGroup_ de@DDExpr{..} t =
+    [PEq (TETypeOfExpr de) (TEExtern gROUP_TYPE [TETVar ddexprCtx (name k), t])]
+    where
+    TypeDef{tdefArgs=[k,_],..} = getType ?d gROUP_TYPE
+
+deIsMap_ :: (?d::DatalogProgram) => DDExpr -> TExpr -> [Predicate]
+deIsMap_ de@DDExpr{..} t =
+    [ PEq (TETypeOfExpr de) (TEExtern mAP_TYPE [TETVar ddexprCtx (name k), TETVar ddexprCtx (name v)])
+    , PEq t (TETuple (Just 2) $ M.fromList [(0, TETVar ddexprCtx (name k)), (1, TETVar ddexprCtx (name v))])]
+    where
+    TypeDef{tdefArgs=[k,v],..} = getType ?d mAP_TYPE
+
+-- Convert type to a type expression, replacing type arguments ('A, 'B, ...)
+-- with type constants 'TETArg "A", TETArg "B", ...'.  For example, when
+-- generating type inference in the body of a function, we treat its type
+-- arguments as constants.  Inferred types for variables and expressions inside
+-- the body of the function may depend on this constants.
```

Comment:
"these" constants.
Could the constants be just other type variables?
This can happen when analyzing a generic function calling another generic function.

Reply:
This is a bit confusing, but when doing type inference inside a function, this functions type arguments are constants in the sense that in the resulting typing variable types can be expressed in terms of these constants and we don't expect to assign concrete types to these constants.  When we call another generic function, it's a completely different situation. That function's type arguments are variables that we actually want to assign to concrete types, and so we model them as variables (TETVar), not constants.